### ISSUE TITLE
Sparse Tensors

### ISSUE THREAD

**Issue by kamathis4** [keras-team/keras-core#670]
Hey @fchollet,

do you have any roadmap to implement sparse tensors for the different backends?
- tf seems to have only one implementation of sparse tensors [COO](https://www.tensorflow.org/guide/sparse_tensor)
- torch seems to have multiple methods but has really good support for [COO](https://pytorch.org/docs/stable/sparse.html#torch-functions-specific-to-sparse-tensors)
- I think for numpy mostly through [scipy](https://docs.scipy.org/doc/scipy/reference/sparse.html)
- JAX has sparse support seems [experimental](https://jax.readthedocs.io/en/latest/jax.experimental.sparse.html) for now

**Comment by fchollet** [comment]
The general plan is to let backend ops receive and return backend-native sparse tensor types. E.g. `backend.tensorflow.numpy.matmul` should be able to receive sparse inputs, in which case it would return sparse inputs.

Then, we passed sparse data input (e.g. via tf.data or via scipy sparse arrays) we would not densify them and just pass them to the backend ops.

Lastly, we would avoid densify gradients in the optimizer (which we currently do).

The most work will be enabling all backend ops with sparse tensors support.

We should do this for the TensorFlow backend first, as it has the most mature support for sparse tensors so far.

**Comment by jackd** [comment]
I'd be willing to do a lot of the dogs-work to make this happen (e.g. write backend wrappers, keras `Operation`s) if some framework for composite tensors (or even just a `SparseTensor` class) could be established.

**Comment by fchollet** [comment]
@hertschuh is currently working on this -- but there may be items you guys can take on!
